Bump Gradle Wrapper to 8.7 and AGP to 8.5.1

AdamGrzybkowski opened the pull request:
## Description:

There's an [issue on Android 15](https://github.com/Yalantis/uCrop/issues/913). UCrop doesn't work well with the enforcement of the edge-to-edge. 

In order to test this using the sample app I need to bump Gradle and AGP. This is what this PR does.

-----

## Test Instructions:

Build the sample app. A successful build should be enough - the sample app doesn't really work well so don't test it (I will fix it in a separate PR).

AdamGrzybkowski committed:
Bump Gradle Wrapper to 8.7 and AGP to 8.5.1

AdamGrzybkowski commented on gradle.properties:
```diff
@@ -18,3 +18,5 @@ org.gradle.jvmargs=-Xmx2048m -XX:MaxMetaspaceSize=512m -XX:+HeapDumpOnOutOfMemor
 
 android.useAndroidX=true
+
+android.nonFinalResIds=false
```
I added it to make as few code changes as possible, though maybe we should handle this properly. Thoughts?

AdamGrzybkowski commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.7-all.zip
```
This is the minimum required to target API 35. Would it be better to use the latest version?

ParaskP7 replied:
Yea, it might be better to target latest Gradle & AGP versions, but then again, I am not 100% sure if it's worth it. Let me think about it and come-back on that for you tomorrow, when I'll do a proper review. 🤔

ParaskP7 replied:
FYI: I just pushed this https://github.com/Automattic/uCrop/pull/10/commits/dc14f64b33792b118fc64ac06f0d0958f2c80185 commit to (properly) update Gradle to [8.7](https://docs.gradle.org/8.7/release-notes.html), which means:

1. Instead of updating the distribution manually, you run this command (`CLI`): `./gradlew wrapper --gradle-version=8.7 --distribution-type=all --gradle-distribution-sha256-sum=194717442575a6f96e1c1befa2c30e9a4fc90f701d7aee33eb879b79e7ff05c0`
2. But, even that is incomplete as running the above command once is not enough. You need to run it twice, once to update the distribution and expected `SHA-256` and again to update the wrapper files themselves ([article](https://sp4ghetticode.medium.com/the-elephant-in-the-room-how-to-update-gradle-in-your-android-project-correctly-09154fe3d47b)). PS: I personally even run it a third time to make sure all Gradle related files are updated.
3. Also, make sure to add `--distribution-type=all` to download the complete (`-all`) ZIP instead of the binary-only (`-bin`) ZIP.
4. Finally, make sure to add `--gradle-distribution-sha256-sum` and point that to the appropriate `distributionSha256Sum`, getting the correct value from [Gradle distribution and wrapper JAR checksum reference](https://gradle.org/release-checksums/).

Hope the above helps! 🙏

ParaskP7 replied:
Also @AdamGrzybkowski, after thinking about it, I think that pointing to Gradle 8.7 and AGP 8.5.1, which is the minimum required to unblock development with API 35, is okay for now. Then, as a separate PR, we could try and update it to the most recent Gradle and AGP versions, just to keep things up-to-date.

AdamGrzybkowski replied:
Thanks @ParaskP7 for the steps! 

About the bump - I agree, it's easy to push everything to one PR, but... This one has a single purpose - unblock the fix so I wanted to focus on that.

ParaskP7 replied:
Awesome @AdamGrzybkowski , and please let me know how to proceed from here, like do you think this is now ready for a(nother) review so I could take a wholistic look at it (again), now that we've settled on what changes we want to merge with this PR? 🙏

AdamGrzybkowski replied:
Yes, please! 🙏 I will go through the changes as well.

AdamGrzybkowski commented:
CI fails with:

```

* What went wrong:
--
  | [2025-02-10T15:31:49Z] An exception occurred applying plugin request [id: 'com.android.application', version: '8.5.1']
  | [2025-02-10T15:31:49Z] > Failed to apply plugin 'com.android.internal.application'.
  | [2025-02-10T15:31:49Z]    > Android Gradle plugin requires Java 17 to run. You are currently using Java 11.
  | [2025-02-10T15:31:49Z]       Your current JDK is located in /usr/lib/jvm/java-11-amazon-corretto.x86_64
  | [2025-02-10T15:31:49Z]       You can try some of the following options:
  | [2025-02-10T15:31:49Z]        - changing the IDE settings.
  | [2025-02-10T15:31:49Z]        - changing the JAVA_HOME environment variable.
  | [2025-02-10T15:31:49Z]        - changing `org.gradle.java.home` in `gradle.properties`.


```


@ParaskP7 I believe this is something that should be set on buildkite, right?

ParaskP7 commented:
> @ParaskP7 I believe this is something that should be set on buildkite, right?

Yea @AdamGrzybkowski this is about [AGP 8.0 and the fact that it requires JDK 17 to run](https://developer.android.com/build/releases/past-releases/agp-8-0-0-release-notes#jdk-17-agp) (also: paqN3M-XM-p2). We've migrated all our project to JDK 17 a long time ago (2 year already) and I never expected we do that again, but yea, this `uCrop` library is currently pointing to [7.4.2](https://developer.android.com/build/releases/past-releases/agp-7-4-0-release-notes), oof, the legacy we need to deal with! 😅 😨 😓

ParaskP7 committed:
Build: Add .java-version to set the project jdk version to 17.0

FYI: jEnv is a tool that lets you manage multiple JDK versions. It has
various features such as setting a global version, a local one, a shell
one and most importantly it supports '.java-version' file and
automatically updates the JDK version if it finds this file. Our CI has
to support multiple JDK versions and as such jEnv support got added,
utilizing '.java-version' files as part of the Android Gradle Plugin
update PRs.

jEnv: https://github.com/jenv/jenv
P2 Post: paqN3M-ZK-p2

ParaskP7 commented:
> Yea @AdamGrzybkowski this is about [AGP 8.0 and the fact that it requires JDK 17 to run](https://developer.android.com/build/releases/past-releases/agp-8-0-0-release-notes#jdk-17-agp) (also: paqN3M-XM-p2). We've migrated all our project to JDK 17 a long time ago (2 year already) and I never expected we do that again, but yea, this uCrop library is currently pointing to [7.4.2](https://developer.android.com/build/releases/past-releases/agp-7-4-0-release-notes), oof, the legacy we need to deal with! 😅 😨 😓

FYI: Just adding the `.java-version` file pointing to JDK `17.0` worked (https://github.com/Automattic/uCrop/pull/10/commits/52733e246fb59a849738238b14f07d76afb4067a). ✅

ParaskP7 committed:
Build: (Properly) Upgrade gradle to 8.7

Release Notes: https://docs.gradle.org/8.7/release-notes.html

Command: ./gradlew wrapper --gradle-version=8.7
--distribution-type=all --gradle-distribution-sha256-sum=
194717442575a6f96e1c1befa2c30e9a4fc90f701d7aee33eb879b79e7ff05c0

ParaskP7 commented:
Btw @AdamGrzybkowski and somehow unrelated, but maybe still related, but after running `./gradlew lint` locally, I found out that we have `6` error (plus `33` warnings) on the `ucrop` modules, and another `2` errors (plus `48` warnings) on the `sample` module. Currently those are being ignored because of the `abortOnError false` on `lintOptions` on both modules, but maybe they shouldn't be. Wdyt? 🤔

PS: Just like I suggested with the Gradle and AGP updates [above](https://github.com/Automattic/uCrop/pull/10#discussion_r1950878421), maybe this is not the right PR to do this work, but instead we should create a separate PR afterwards to make sure that (at least) all those Lint errors are resolved, just to make sure that when other project are using this library there are not going to be surprised by an associated issue.

ParaskP7 commented on ucrop/build.gradle:
```diff
@@ -9,6 +9,7 @@ repositories {
 }
 
 android {
+    namespace "com.yalantis.ucrop"
```
Minor (🔍): Now that this is defined here the whole [AndroidManifest.xml](https://github.com/Automattic/uCrop/blob/develop-non-native/ucrop/src/main/AndroidManifest.xml) file could be deleted from within the `ucrop` module.

AdamGrzybkowski replied:
TIL - I thought `AndroidManifest.xml` presence is required even without config.

ParaskP7 replied:
Yea, no, bye-bye XML and thanks for all the fish (pain)! 🤪

ParaskP7 commented on sample/build.gradle:
```diff
@@ -8,6 +8,7 @@ repositories {
 }
 
 android {
+    namespace "com.yalantis.ucrop.sample"
```
Minor (🔍): Now that this is defined here, this [package="com.yalantis.ucrop.sample"](https://github.com/Automattic/uCrop/blob/develop-non-native/sample/src/main/AndroidManifest.xml#L3) definition could be deleted from the [AndroidManifest.xml](https://github.com/Automattic/uCrop/blob/develop-non-native/sample/src/main/AndroidManifest.xml) file within the `sample` module.

ParaskP7 commented:
> [Yes, please! 🙏 I will go through the changes as well.](https://github.com/Automattic/uCrop/pull/10#discussion_r1954200023)

Deal @AdamGrzybkowski , I did leave a couple new comments but as soon as you go through my changes too, then give me a 👍 (or not), I'll do a final review and give you a ✅ on this. 👍

AdamGrzybkowski committed:
Remove package from Sample app's AndroidManifest.xml

AdamGrzybkowski committed:
Remove ucrop/src/main/AndroidManifest.xml

It's empty now so can be safely removed.

AdamGrzybkowski commented:
LGTM @ParaskP7

ParaskP7 reviewed (approved):
LGTM too @AdamGrzybkowski ! 🚀
